### PR TITLE
[feature-wip](array-type) Support loading data in vectorized format

### LINKED ISSUE #10064: [Bug][array-type] Failed to select array type data after enable storage vectorization.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master

### What's Wrong?

In commit #9848 , the configuration `enable_storage_vectorization` in `BE` was enabled by default.

However, the `ArrayFileColumnIterator` didn't implement the interface `Status next_batch(size_t* n, vectorized::MutableColumnPtr& dst, bool* has_null)`. Thereby, the array type data can't be retrieved from underlying data file.

### What You Expected?

Select array type data successfully.

### How to Reproduce?

```shell
mysql> create table array_type_table (k1 INT, k2 Array<int>) duplicate key (k1) 
    -> distributed by hash(k1) buckets 1 properties('replication_num' = '1');
mysql> set enable_array_type = true;
mysql> set enable_vectorized_engine = true;
mysql> insert into array_type_table values (0, []), (1, [NULL]), (2, [1, 2, 3]), (3, [1, NULL, 3]); 
mysql> select * from array_type_table;
Empty set (0.03 sec)
```

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by adonis0147**
# Proposed changes

Issue Number: close #10064

## Problem Summary:

Please refer to #10064

## Checklist(Required)

1. Does it affect the original behavior: No
2. Has unit tests been added: Yes
3. Has document been added or modified: No Need
4. Does it need to update dependencies: No
5. Are there any changes that cannot be rolled back: No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 1]
```diff
@@ -461,6 +461,46 @@ Status ArrayFileColumnIterator::next_batch(size_t* n, ColumnBlockView* dst, bool
...
+Status ArrayFileColumnIterator::next_batch(size_t* n, vectorized::MutableColumnPtr& dst,
+                                           bool* has_null) {
+    const auto* column_array = vectorized::check_and_get_column<vectorized::ColumnArray>(
+            dst->is_nullable() ? static_cast<vectorized::ColumnNullable&>(*dst).get_nested_column()
+                               : *dst);
+
+    bool offsets_has_null = false;
+    auto column_offsets_ptr = column_array->get_offsets_column().assume_mutable();
+    auto start = column_offsets_ptr->size();
+    _length_iterator->next_batch(n, column_offsets_ptr, &offsets_has_null);
```
Comment (cambyzju) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:473]: RETURN_IF_ERROR

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 2]
```diff
@@ -461,6 +461,46 @@ Status ArrayFileColumnIterator::next_batch(size_t* n, ColumnBlockView* dst, bool
...
+    auto& column_offsets =
+            static_cast<vectorized::ColumnArray::ColumnOffsets&>(*column_offsets_ptr);
+    auto& offsets_data = column_offsets.get_data();
+    for (int i = start; i < offsets_data.size(); ++i) {
```
Comment (cambyzju) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:477]:
```suggestion
    for (size_t i = start; i < offsets_data.size(); ++i) {
```
we will change offsets of array to int64 later.

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 3]
```diff
@@ -461,6 +461,46 @@ Status ArrayFileColumnIterator::next_batch(size_t* n, ColumnBlockView* dst, bool
...
+    auto column_items_ptr = column_array->get_data().assume_mutable();
+    for (int64_t i = start; i < offsets_data.size(); ++i) {
+        auto offset = offsets_data[i - 1]; // -1 is ok.
+        auto size = offsets_data[i] - offset;
+        if (size == 0) {
+            continue;
+        }
+        size_t num_items = size;
+        bool items_has_null = false;
+        _item_iterator->next_batch(&num_items, column_items_ptr, &items_has_null);
```
Comment (cambyzju) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:490]: First get total item size, then call next_batch repeatedly until reach total item size, we could reduce the call of next_batch.

File: be/src/olap/rowset/segment_v2/column_reader.cpp [hunk 4]
```diff
@@ -461,6 +461,46 @@ Status ArrayFileColumnIterator::next_batch(size_t* n, ColumnBlockView* dst, bool
...
+    bool offsets_has_null = false;
+    auto column_offsets_ptr = column_array->get_offsets_column().assume_mutable();
+    auto start = column_offsets_ptr->size();
+    _length_iterator->next_batch(n, column_offsets_ptr, &offsets_has_null);
+    auto& column_offsets =
+            static_cast<vectorized::ColumnArray::ColumnOffsets&>(*column_offsets_ptr);
+    auto& offsets_data = column_offsets.get_data();
+    for (int i = start; i < offsets_data.size(); ++i) {
+        offsets_data[i] += offsets_data[i - 1]; // -1 is ok
+    }
```
Comment (cambyzju) [inline be/src/olap/rowset/segment_v2/column_reader.cpp:479]: we need set `n` after read data success.

**Comment by eldenmoon** [comment]
Sorry for my issue in this PR.

**Comment by eldenmoon** [comment]
Sorry for my issue in this PR, I'm currently using Array, found that when  BE build with ASAN,  the column type is like ARRAY<STRING>,  there will be a coredump which is heap-use-after-free after I insert a data with `enable_vectorized_engine`

**Comment by eldenmoon** [comment]
Sorry for my issue in this PR
I'm currently using Array, found that when I build BE with ASAN,  using the column type  like `ARRAY<STRING>`, after I insert with vectorized engine enabled, there will be a coredump which is heap-use-after-free, see issue  #9934  for more details. I commented the clue I found in this PR https://github.com/apache/incubator-doris/pull/9466 (the last comment) , but I'm not quite sure how to solve it

**Comment by adonis0147** [comment]
Hi @eldenmoon , thanks for your feedback. I will investigate the issue you mentioned and try to solve it.

File: be/src/olap/rowset/segment_v2/column_reader.cpp (hunk 1 above)
Comment (adonis0147) [reply be/src/olap/rowset/segment_v2/column_reader.cpp]: Added

File: be/src/olap/rowset/segment_v2/column_reader.cpp (hunk 3 above)
Comment (adonis0147) [reply be/src/olap/rowset/segment_v2/column_reader.cpp]: Refined.

File: be/src/olap/rowset/segment_v2/column_reader.cpp (hunk 4 above)
Comment (adonis0147) [reply be/src/olap/rowset/segment_v2/column_reader.cpp:482]: The value of `*n` was updated at [L413](https://github.com/apache/incubator-doris/blob/c4bdfb2d1194f16b923837785719a0edb2eaf78c/be/src/olap/rowset/segment_v2/column_reader.cpp#L413)

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**